test/controllers/tasks_controller_test.rb

```diff
@@ -77,52 +77,170 @@
+      }.wont_change "Task.count"
+
+      must_respond_with :redirect
+
+      task = Task.find_by(id: id)
+      expect(task.task_name).must_equal new_task_hash[:task][:task_name]
+      expect(task.description).must_equal new_task_hash[:task][:description]
+      expect(task.date_due).must_equal new_task_hash[:task][:date_due]
+    end
+
+    # THIS IS THE FIRST THING I TRIED WRITING. IT GAVE AN ERROR.
+    # ActiveModel::UnknownAttributeError: unknown attribute 'params' for Task.
     it "can update an existing task" do
       skip
-      # Your code here
+      # Arrange
+      change_task = Task.find_by(task_name: "Clean the bathroom")
+
+      change_task.update(params: new_task_hash)
+
+      # Assert
+      expect { }.must_change
+      must_respond_with :success
+      must_redirect_to task_path
     end
 
     it "will redirect to the root page if given an invalid id" do
-      skip
-      # Your code here
+      # skip
+      # Act
+      get task_path(-1)
+
+      # Assert
+      must_respond_with :redirect
     end
   end
 
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    before do
+      Task.create(task_name: "Acquire a rain jacket", description: "One that is waterproof; check The Bins.", date_due: "2020-03-15")
 
+      Task.create(task_name: "Take a break!", description: "Go outside, play your cello, take your eyes away from the screen. Do it!!", date_due: "2020-03-15")
+    end
+
+    it "will destroy an existing task" do
+      # skip
+      # Arrange
+      task_to_delete = Task.find_by(task_name: "Acquire a rain jacket")
+
+      other_task = Task.find_by(task_name: "Take a break!")
+
+      # task_to_delete.destroy
+
+      # Assert
+
+      # NEED HELP WITH THE NEXT THREE TESTS!!!
+      expect {
+        task_to_delete.destroy
+      }.must_change "Task.count", -1
+      # must_respond_with :success
+
+      # expect {
+      #   other_task.destroy
+      # }.must_respond_with :success
+      # must_redirect_to tasks_path
+    end
+
+    # it "will give a 404 error if trying to delete a non-existant task" do
+    #   task_to_delete = Task.find_by(id: -1)
+    #   expect {
+    #     task_to_delete.destroy
+    #   }.must_respond_with :missing
+    # end
   end
 
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    # Arrange
+    before do
+      Task.create(task_name: "Finish writing this test.", description: "One that is waterproof; check The Bins.", date_due: "2020-03-15")
+
+      Task.create(task_name: "Take another break!", description: "For real this time.", date_due: "2019-04-14", completed: true)
+    end
+
+    task_to_toggle1 = Task.find_by(task_name: "Finish writing this test.")
+    task_to_toggle2 = Task.find_by(task_name: "Take another break!")
+    # expect {
+    #   task_to_toggle1.toggle_completed
+    # }.must_change task_to_toggle[:completed], true
+
+    # Act
+    task_to_toggle2.toggle_completed
```

Comment:
Here, you are calling `toggle_completed` as if it were an instance method on the model `task_to_toggle2`. However, your model doesn't have a method by that name defined! Instead, your Act will be to execute the code in the `TasksController` in the action/method named `toggle_compelted`... And how do you get to that code? Through the router! Your act will be whatever is the route defined to get there... which will be `put toggle_completed_action_path`

I think fixing the rest of your incomplete tests will follow this same pattern:
1. Determine which controller and controller action you are testing
1. Determine which route will get to that controller and controller action (`delete task_path`, maybe?)
1. Determine which info needs to go along with it:
    - Does it have route params?
    - Does it have form data?
1. Write out the "Act" in this test using that route and that additional information